src/cpic/model/board_state.h

```diff
@@ -65,69 +59,82 @@ class BoardColumn {
 class BoardState {
 public:
   BoardState() = default;
-  explicit BoardState(std::vector<BoardColumn> table) : internal(std::move(table)) {}
+  explicit BoardState(const std::vector<std::vector<Color>> &rows);
 
   bool isValid(const Board *) const;
   unsigned short countColorInColumn(unsigned short, Color) const;
   unsigned short countColorInRow(unsigned short, Color) const;
+  short findFirstInRow(unsigned short, Color) const;
+  short findLastInRow(unsigned short, Color) const;
 
-  void push_back(const BoardColumn &row) {
-    internal.push_back(row);
-  }
-
-  void setColorAt(unsigned short column, unsigned short row, Color color) {
-    internal.at(column).setColorAt(row, color);
-  }
-
-  unsigned long long int rowCount() const {
-    return internal[0].rowCount();
-  }
-
-  unsigned long long int columnCount() const {
-    return internal.size();
+  inline void setColorAt(unsigned short column, unsigned short row, Color color) {
+    rows.at(row).setColorAt(column, color);
   }
 
-  const std::vector<BoardColumn>::const_iterator begin() const {
-    return internal.begin();
+  inline unsigned long long int rowCount() const {
+    return rows.size();
   }
 
-  const std::vector<BoardColumn>::const_iterator end() const {
-    return internal.end();
+  inline unsigned long long int columnCount() const {
+    return rows[0].columnCount();
   }
 
-  Color colorAt(unsigned short column, unsigned short row) const {
-    return internal.at(column).row(row);
+  inline Color colorAt(unsigned short column, unsigned short row) const {
+    return rows.at(row).column(column);
   }
 
   bool operator==(const BoardState &other) const {
-    return internal == other.internal;
+    return rows == other.rows;
   }
 
   bool operator!=(const BoardState &other) const {
-    return internal != other.internal;
+    return rows != other.rows;
   }
 
 private:
-  std::vector<BoardColumn> internal;
+  std::vector<BoardRow> rows;
 };
 
 inline BoardState pivotState(std::vector<std::vector<Color>> table) {
-  auto rowCount = static_cast<unsigned short>(table.size());
-  auto columnCount = static_cast<unsigned short>(table[0].size());
-  std::vector<BoardColumn> result(columnCount, BoardColumn(rowCount));
-
-  for (unsigned short i = 0; i < rowCount; ++i) {
-    for (unsigned short j = 0; j < columnCount; ++j) {
-      result[i].setColorAt(j, table[j][i]);
-    }
+  return BoardState(table);
+}
+
+inline BoardState emptyBoardState(ushort columnCount, ushort rowCount) {
+  //FIXME: Implement this function properly
```

Comment:
Maybe this FIXME should actually say that a BoardState by default should be created as empty?